nexus/src/app/background/tasks/saga_recovery.rs

```diff
@@ -0,0 +1,916 @@
+        else {
+            panic!("saga recovery failed");
+        };
+
+        assert_eq!(last_pass_success.nrecovered, 0);
+        assert_eq!(last_pass_success.nfailed, 0);
+        assert_eq!(last_pass_success.nskipped, 0);
+
+        // The nodes should not have been replayed.
+        assert_eq!(uctx.n1_count.load(Ordering::SeqCst), 1);
+        assert_eq!(uctx.n2_count.load(Ordering::SeqCst), 1);
+
+        // Test cleanup
+        drop(task);
+        let sec_client = Arc::try_unwrap(sec_client).unwrap();
+        sec_client.shutdown().await;
+        db.cleanup().await.unwrap();
+        logctx.cleanup_successful();
+    }
+
+    // Verify the plumbing that exists between regular saga creation and saga
+    // recovery.
+    #[nexus_test(server = crate::Server)]
+    async fn test_nexus_recovery(cptestctx: &ControlPlaneTestContext) {
+        let nexus = &cptestctx.server.server_context().nexus;
+
+        // This is tricky to do.  We're trying to make sure the plumbing is
+        // hooked up so that when a saga is created, the saga recovery task
+        // learns about it.  The purpose of that plumbing is to ensure that we
+        // don't try to recover a task that's already running.  It'd be ideal to
+        // test that directly, but we can't easily control execution well enough
+        // to ensure that the background task runs while the saga is still
+        // running.  However, even if we miss it (i.e., the background task only
+        // runs after the saga completes successfully), there's a side effect we
+        // can look for: the task should report the completed saga as "maybe
+        // done".  On the next activation, it should report that it's removed a
+        // saga from its internal state (because it saw that it was done).
+
+        // Wait for the task to run once.
+        let driver = nexus.background_tasks_driver.get().unwrap();
+        let task_name = driver
+            .tasks()
+            .find(|task_name| task_name.as_str() == "saga_recovery")
+            .expect("expected background task called \"saga_recovery\"");
+        let first_completed = wait_for_condition(
+            || async {
+                let status = driver.task_status(task_name);
+                let LastResult::Completed(completed) = status.last else {
+                    return Err(CondCheckError::<()>::NotYet);
+                };
+                Ok(completed)
+            },
+            &std::time::Duration::from_millis(250),
+            &std::time::Duration::from_secs(15),
+        )
+        .await
+        .unwrap();
+
+        // Make sure that it didn't find anything to do.
+        let status_raw = first_completed.details;
+        let status: nexus_saga_recovery::Report =
+            serde_json::from_value(status_raw).unwrap();
+        let nexus_saga_recovery::LastPass::Success(last_pass_success) =
+            status.last_pass
+        else {
+            panic!("wrong last pass variant");
```

Comment:
teeny nitpick: should we format what we actually did get in the panic message, here?

Reply:
Is there an easy way to do that with `let ... else`?

Reply:
no, i think you would need to use a `match` instead. it's not that important...